polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -126,6 +133,229 @@ void createLocalBarrier(OpBuilder builder) {
+
+  // Collect the global thread ids used in the function the loop is in.
+  auto funcOp = loop->template getParentOfType<FunctionOpInterface>();
+  SmallVector<Value> threadVars =
+      memAccessAnalysis.computeThreadVector(funcOp, solver);
+
+  // Collect candidate memref accesses in the loop.
+  loop->walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (!isa<affine::AffineLoadOp, affine::AffineStoreOp>(op))
+      return;
+
+    affine::MemRefAccess access(op);
+    addMemRefAccess(access);
+  });
+
+  // Analyze the accesses collected and populate the 'accessToMemSpace' map.
+  for (auto &entry : accesses) {
+    ArrayRef<affine::MemRefAccess> accesses = entry.second;
+
+    // Filter out memRef accesses that aren't read-only.
+    if (CollectReadOnlyAccesses && !areReadOnly(accesses))
+      continue;
+
+    bool temporalReuse = haveTemporalReuse(accesses, threadVars);
+
+    for (const affine::MemRefAccess &access : accesses) {
+      LLVM_DEBUG(llvm::dbgs() << "Classify: " << *access.opInst << "\n");
+
+      std::optional<MemoryAccess> memAccess =
+          memAccessAnalysis.getMemoryAccess(access);
+      if (!memAccess.has_value()) {
+        LLVM_DEBUG(llvm::dbgs() << "Unable to analyze memory access\n");
+        continue;
+      }
+
+      // Get the inter-thread access pattern and classify the memory access.
+      MemoryAccessMatrix interThreadMatrix =
+          memAccess->getInterThreadAccessMatrix(threadVars.size());
+      MemoryAccessPattern interThreadAccessPattern = MemoryAccess::classify(
+          interThreadMatrix, memAccess->getOffsetVector(), solver);
+
+      switch (interThreadAccessPattern) {
+      case Linear:
+      case Reverse:
+      case ReverseLinear:
+        // These patterns lead to fully coalesced memory accesses.
+        accessToMemSpace[&access] = MemorySpace::Global;
+        break;
+      case Shifted:
+      case LinearShifted:
+      case ReverseLinearShifted:
+      case LinearOverlapped:
+      case ReverseLinearOverlapped:
+        // These patterns lead to partially coalesced memory accesses.
+        accessToMemSpace[&access] = MemorySpace::Global;
+        break;
+      case Strided:
+      case ReverseStrided:
+      case StridedShifted:
+      case ReverseStridedShifted:
+      case Overlapped:
+      case StridedOverlapped:
+      case ReverseStridedOverlapped: {
+        Value strideVal =
+            interThreadMatrix(interThreadMatrix.getNumRows() - 1,
+                              interThreadMatrix.getNumColumns() - 1);
+
+        // Use shared memory only if the stride is known and it's greater than a
+        // sufficiently large value (small stride values yield partially
+        // coalesed memory accesses).
+        if (auto stride = getConstIntegerValue(strideVal, solver)) {
+          bool strideIsLargeEnough = stride->sgt(8) || stride->slt(-8);
+          bool useSharedMemory =
+              temporalReuse && (stride->isZero() || strideIsLargeEnough);
```

Comment:
Do you think `stride->isZero()` should be part of `strideIsLargeEnough`?

Reply:
Not really because a zero stride  names "large enough" is an oxymoron. The reason that for a zero stride we want to use shared memory is that with a zero stride the access pattern would skip an entire row.   I will improve the comment.

Reply:
done